fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -0,0 +1,154 @@
+use std::time::Duration;
+
+use fedimint_core::db::Database;
+use fedimint_core::task::sleep;
+use fedimint_core::PeerId;
+use tokio::sync::{mpsc, oneshot, watch};
+
+use crate::keychain::Keychain;
+use crate::{
+    db, session, AcceptedItem, Decision, Message, OrderedItem, Recipient, Shutdown, SignedBlock,
+};
+
+async fn relay_messages(
+    db: Database,
+    incoming_message_receiver: async_channel::Receiver<(Message, PeerId)>,
+    outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    network_data_sender: async_channel::Sender<Vec<u8>>,
+    signed_block_sender: async_channel::Sender<SignedBlock>,
+) {
+    while let Ok((message, peer_id)) = incoming_message_receiver.recv().await {
+        match message {
+            Message::NetworkData(network_data) => {
+                network_data_sender.send(network_data).await.ok();
+            }
+            Message::BlockRequest(index) => {
+                if let Some(signed_block) = db::load_block(&db, index).await {
+                    outgoing_message_sender
+                        .send((Message::Block(signed_block), Recipient::Peer(peer_id)))
+                        .await
+                        .ok();
+
+                    tracing::info!("Served the block with index {} to peer {}", index, peer_id);
+                }
+            }
+            Message::Block(signed_block) => {
+                signed_block_sender.send(signed_block).await.ok();
+            }
+        }
+    }
+
+    std::future::pending().await
+}
+
+async fn process_signed_block(
+    signed_block: SignedBlock,
+    ordered_item_sender: mpsc::Sender<OrderedItem>,
+) -> anyhow::Result<()> {
+    for AcceptedItem { item, peer_id } in signed_block.block.items.iter() {
+        let (decision_sender, decision_receiver) = oneshot::channel();
+
+        ordered_item_sender.try_send(OrderedItem {
+            item: item.clone(),
+            block_index: signed_block.block.index,
+            peer_id: *peer_id,
+            decision_sender,
+        })?;
+
+        // The threshold signed blocks items have to be accepted by Fedimint Consensus.
+        assert!(decision_receiver.await? == Decision::Accept);
```

Comment:
Is there a reason not to send all items first, collect the receivers, and only then assert responses? That would eliminate the ping-pong-latency here, afaict.

Also - it looks a bit odd altogether, as I don't understand why would items be sent one by one, instead of once in some vector. (but I'm scanning whole PR top to bottom, so maybe will understand it later).

Reply:
Well this function models the behaviour of a running session sending ordered items; in the latter case it is a little simpler to do this ping pong style. However, I changed it to use vectors as well